Add ramped charge value for Tesla

dalathegreat opened the pull request:
### What
This PR adds a new feature for Tesla batteries, smoothing of charge max value for when battery approaches 100%

### Why
@Newevg noticed that when trying to charge their LFP Tesla battery to 100%, charging it with 15kW all the way up to 100% was not possible. To reach 100%, the charge needs to taper off, otherwise the contactors would open, and the BMS would flag a fault. 

### How
The .header file now has a configurable max value, `MAXCHARGEPOWERALLOWED`. This value, which defaults to 15000W, will be used when battery is between 0-95%. When battery reaches 95%, it ramps between `MAXCHARGEPOWERALLOWED`-> 0 in the 95-100% range. This should allow the battery to more gracefully reach 100%.

dalathegreat commented on Software/src/battery/TESLA-MODEL-3-BATTERY.cpp:
```diff
@@ -196,9 +196,11 @@ void update_values_tesla_model_3_battery() {  //This function maps all the value
-  } else {
-    max_target_charge_power = 15000;  //Otherwise we can push 15kW into the pack!
+    max_target_charge_power = 0;  // When battery is 100% full, set allowed charge W to 0
+  } else if (SOC > 9500) {        // When battery is between 95-99.99%, ramp the value between Max<->0
+    max_target_charge_power = MAXCHARGEPOWERALLOWED * (1 - (SOC - 9500) / 500.0);
```
Max cell voltage will be proportional to charge speed, so basing the ramp on max cell voltage might cause oscillations.

I instead propose we use SOC_VI for the ramp. That way incase user limits battery to between 20-80%, the value will never ramp. And incase user extends the range closer to 0-100%, we will use a ramped value at the extremes.

matvey83 replied:
OK makes sense, basing the ramp on the real SOC is the way to go.  Also how is the 95% number derived?  Seems like this number will change based the max charge power and battery capacity.  So for some configurations we may want to start ramping at 90% for example.

dalathegreat replied:
@matvey83 The 95% number comes from watching multiple different battery vendors throttle charging at end of SOC%. Fastcharging usually starts throttling at 80%, but AC slowcharging is ~95% where it starts to taper off.

matvey83 replied:
Thanks

dalathegreat commented:
@matvey83 and @lenvm , I reworked the ramp logic to avoid having users that restrict their battery encounter slow charging.

I also added an extra safety check in bf292a1

dalathegreat commented on Software/src/battery/TESLA-MODEL-3-BATTERY.cpp:
```diff
@@ -156,7 +156,7 @@ static const char* hvilStatusState[] = {"NOT OK",
 
-#define MAX_CELL_VOLTAGE_LFP 3450   //Battery is put into emergency stop if one cell goes over this value
+#define MAX_CELL_VOLTAGE_LFP 3500   //Battery is put into emergency stop if one cell goes over this value
 #define MIN_CELL_VOLTAGE_LFP 2800   //Battery is put into emergency stop if one cell goes over this value
```
Fixed in 6801e83

lenvm commented on Software/src/battery/TESLA-MODEL-3-BATTERY.cpp:
```diff
@@ -193,12 +193,17 @@ void update_values_tesla_model_3_battery() {  //This function maps all the value
   } else {
     max_target_discharge_power = temporaryvariable;
   }
+  if (SOC < 20) {  // When battery is lower than 0.20% , set allowed discharge W to 0
```
Instead of hardcoding the value 20, shouldn’t this be based on MINPERCENTAGE?

Minor comment: the comment 0.20% is not correct.

lenvm replied:
I would also change these lines into one `if, else if, else` statement that defines the `max_target_discharge_power`, rather than two separate `if` statements.

dalathegreat replied:
@lenvm fixed in e2d3c4f , I decided against using MINPERCENTAGE, and going for the simpler if SOC=0 check, since that is what other battery types use! Nice bonus with the temporaryvariable removed!

lenvm commented on Software/src/battery/TESLA-MODEL-3-BATTERY.cpp:
```diff
@@ -193,12 +193,17 @@ void update_values_tesla_model_3_battery() {  //This function maps all the value
+  if (soc_vi > 950) {  // When battery is between 95-99.99% real SOC, ramp the value between Max<->0
+    max_target_charge_power = MAXCHARGEPOWERALLOWED * (1 - (soc_vi - 950) / 50.0);
+  }
+  if (SOC == 10000) {  // When battery is defined as 100% full, set allowed charge W to 0
```
instead of two separate if-statements, I suggest using one large `if, else if, else` statement:
```
if (SOC == 10000) {
  max_target_charge_power = 0;
} else if (soc_vi > 950) {
  max_target_charge power = ...;
} else {
  max_target_charge_power = MAXCHARGEPOWERALLOWED;
}
```

dalathegreat replied:
Good improvement, fixed in 36deffd

lenvm commented:
Another way to implement the charge power limit, is to use the charge limit that may be reported by the battery.

According to [this repository](https://github.com/joshwardell/model3dbc/blob/master/Model3CAN.dbc), messages [594 = 0x252](https://github.com/joshwardell/model3dbc/blob/c56c113c353d09b1a5dd57677461f02fe8b7c6ca/Model3CAN.dbc#L2628-L2634) and [722 = 0xD2D](https://github.com/joshwardell/model3dbc/blob/c56c113c353d09b1a5dd57677461f02fe8b7c6ca/Model3CAN.dbc#L2381-L2385) could report the BMS charge and discharge limits.

Also the following messages could be interesting:
- [637 = 0x27D](https://github.com/joshwardell/model3dbc/blob/c56c113c353d09b1a5dd57677461f02fe8b7c6ca/Model3CAN.dbc#L487-L492)
- [701 = 0x2BD](https://github.com/joshwardell/model3dbc/blob/c56c113c353d09b1a5dd57677461f02fe8b7c6ca/Model3CAN.dbc#L494-L496)

Are any of these messages available (in logs) on the Tesla Battery CAN?

dalathegreat commented:
@lenvm For some reason the Tesla packs are not satisfied, and will show 0kW on `regenerative_limit`
![bild](https://github.com/dalathegreat/BYD-Battery-Emulator-For-Gen24/assets/26695010/7d3ee12c-32b9-4e58-85e9-5064bd6155ea)

So this is why we need the ramp, for now!
